theano/gpuarray/c_code/dnn_conv_base.c

```diff
@@ -33,3 +56,205 @@ if (APPLY_SPECIFIC(output) != NULL)
+{
+    std::ostringstream s;
+    if (size > 0) {
+      
+      s << res[0];
+      for (int i = 1; i < size; ++i)
+        s <<',' << res[i];
+    }
+    return s.str();
+}
+
+static std::string shape(cudnnTensorDescriptor_t t)
+{
+    // cuDNN can handle up to CUDNN_DIM_MAX dimensions.
+    int res[CUDNN_DIM_MAX];
+    int stride[CUDNN_DIM_MAX];
+    int nbDims;
+    cudnnDataType_t type;
+    checkCudnnStatus(cudnnGetTensorNdDescriptor(t, CUDNN_DIM_MAX, &type, &nbDims, res, stride),
+                     "error getting tensor description");
+    if (PyErr_Occurred()) return "";
+    return shape(res, nbDims) + "," + shape(stride, nbDims);
+};
+
+static std::string shape(cudnnFilterDescriptor_t t, cudnnDataType_t* type)
+{
+    cudnnTensorFormat_t format;
+    int res[CUDNN_DIM_MAX];
+    int outDims;
+    checkCudnnStatus(cudnnGetFilterNdDescriptor(t, CUDNN_DIM_MAX, type, &format, &outDims, res),
+                     "error getting filter description");
+    if (PyErr_Occurred()) return "";
+    return shape(res, outDims);
+};
+
+static std::string shape(cudnnConvolutionDescriptor_t convDesc)
+{
+    int nDim;
+    cudnnConvolutionMode_t mode;
+    cudnnDataType_t        computeType;
+    
+    int                                 padA[5];
+    int                                 strideA[5];
+    int                                 dilationA[5];
+
+    checkCudnnStatus(
+        cudnnGetConvolutionNdDescriptor( convDesc, 5,
+                                         &nDim,
+                                         &padA[0],
+                                         &strideA[0],
+                                         &dilationA[0],
+                                         &mode,
+                                         &computeType ),
+        "error getting convolution description");
+    if (PyErr_Occurred()) return "";
+    
+    return (std::string("-mode ") +
+            ((mode == CUDNN_CONVOLUTION) ? "conv" : "cross") +
+            " -pad " +
+            shape(padA, nDim) +
+            " -subsample " +
+            shape(strideA, nDim) +
+            " -dilation " +
+            shape(dilationA, nDim));
+}
+
+static bool all_aligned(cudnnDataType_t type, void* in, void* out, void* filter)
+{
+        size_t alignMask = (type == CUDNN_DATA_HALF) ? 0x7F : 0xFF ;
+        // there have to be entries for both aligned and not
+        if (((size_t)in | (size_t)out | (size_t)filter) & alignMask)
+        {
+            return false;
+        }
+        return true;
+}
+
+static std::string dnn_conv_shape(cudnnTensorDescriptor_t inputDesc, PyGpuArrayObject* input,
+				  cudnnFilterDescriptor_t filterDesc, PyGpuArrayObject* filter,
+				  cudnnConvolutionDescriptor_t convDesc,
+				  PyGpuArrayObject* output, int groups)
+{
+    cudnnDataType_t  dType;
+    std::ostringstream s;
+    int expected_output_dims[5] = {0};
+    cudnnStatus_t err = cudnnGetConvolutionNdForwardOutputDim(convDesc, inputDesc, filterDesc,
```

Comment:
There are 2 Jenkins tests that will still fail:
- http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/672/testReport/junit/theano.gpuarray.tests.test_abstractconv/TestDnnConv2d/test_gradinput_impossible_output_shapes__1__1__1__1____1__1__2__2___2___valid___1_/
- http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/672/testReport/junit/theano.gpuarray.tests.test_abstractconv/TestDnnConv2d/test_gradinput_impossible_output_shapes__1__1__1__1____1__1__2__2___3___valid___1_/

These tests expect cuDNN to fail for inputs/filters shapes that should produce output shapes assumed to be bad. They currently fail with default algorithms (none for dnn_gi and dnn_gw, small for dnn_fwd), but they succeed when I test with runtime-chosen algorithms (e.g. time_once). I think it's because output checking that occurs here is now executed only for runtime-chosen algorithms, as `dnn_conv_shape()` is called only in such cases.

I will put the output checking to an independent function and ensure it is always called.